File: 07-mobile-native/03-android-java/app/src/main/java/com/example/todolist/model/Todo.java
package com.example.todolist.model;

import androidx.room.ColumnInfo;
import androidx.room.Entity;
import androidx.room.PrimaryKey;

import java.util.Objects;

/**
 * Todo Entity for Room Database
 *
 * Represents a todo item in the SQLite database.
 * Uses Room annotations for ORM mapping.
 */
@Entity(tableName = "todos")
public class Todo {

    @PrimaryKey(autoGenerate = true)
    @ColumnInfo(name = "id")
    private long id;

    @ColumnInfo(name = "text")
    private String text;

    @ColumnInfo(name = "completed")
    private boolean completed;

    @ColumnInfo(name = "created_at")
    private long createdAt;

    /**
     * Default constructor (required by Room)
     */
    public Todo() {
        this.completed = false;
        this.createdAt = System.currentTimeMillis();
    }

    /**
     * Constructor with text
     *
     * @param text the todo text
     */
    public Todo(String text) {
        this.text = text;
        this.completed = false;
        this.createdAt = System.currentTimeMillis();
    }

    /**
     * Full constructor
     *
     * @param id        the todo ID
     * @param text      the todo text
     * @param completed the completion status
     * @param createdAt the creation timestamp
     */
    public Todo(long id, String text, boolean completed, long createdAt) {
        this.id = id;
        this.text = text;
        this.completed = completed;
        this.createdAt = createdAt;
    }

    // ==================== Getters and Setters ====================

    public long getId() {
        return id;
    }

    public void setId(long id) {
        this.id = id;
    }

    public String getText() {
        return text;
    }

    public void setText(String text) {
        this.text = text;
    }

    public boolean isCompleted() {
        return completed;
    }

    public void setCompleted(boolean completed) {
        this.completed = completed;
    }

    public long getCreatedAt() {
        return createdAt;
    }

    public void setCreatedAt(long createdAt) {
        this.createdAt = createdAt;
    }

    // ==================== Helper Methods ====================

    /**
     * Toggle the completion status
     */
    public void toggle() {
        this.completed = !this.completed;
    }

    /**
     * Get formatted creation date/time
     *
     * @return formatted date string
     */
    public String getFormattedCreatedAt() {
        java.text.SimpleDateFormat sdf = new java.text.SimpleDateFormat("yyyy-MM-dd HH:mm", java.util.Locale.getDefault());
        return sdf.format(new java.util.Date(createdAt));
    }

    /**
     * Get status display text
     *
     * @return "Completed" or "Active"
     */
    public String getStatusText() {
        return completed ? "Completed" : "Active";
    }

    // ==================== Object Methods ====================

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Todo todo = (Todo) o;
        return id == todo.id;
    }

    @Override
    public int hashCode() {
        return Objects.hash(id);
    }

    @Override
    public String toString() {
        return String.format("Todo[id=%d, text='%s', completed=%s, createdAt=%s]",
                id, text, completed, getFormattedCreatedAt());
    }
}

File: 07-mobile-native/02-android-compose/app/src/main/java/com/example/todocompose/data/Todo.kt
package com.example.todocompose.data

import androidx.compose.runtime.Immutable
import androidx.room.Entity
import androidx.room.PrimaryKey
import kotlinx.serialization.Serializable

/**
 * Todo - Data class representing a single todo item
 *
 * This immutable data class follows Kotlin best practices and serves as the
 * core domain model for the application. It uses:
 * - Room annotations for database persistence
 * - kotlinx.serialization for JSON serialization
 * - @Immutable for Compose optimization
 *
 * Key Design Decisions:
 * 1. Immutable (val properties + @Immutable): Ensures thread safety and Compose optimization
 * 2. Data class: Automatically provides equals(), hashCode(), copy(), toString()
 * 3. Room Entity: Database table representation
 * 4. Serializable: Enables JSON conversion for DataStore/backup
 * 5. UUID for IDs: Ensures globally unique identifiers
 * 6. Timestamp: Tracks creation time for sorting/filtering
 *
 * @property id Unique identifier for the todo item (UUID string)
 * @property text The todo description/title
 * @property completed Whether the todo is marked as done
 * @property timestamp Creation time in milliseconds since epoch
 *
 * Example usage:
 * ```kotlin
 * val todo = Todo(
 *     id = UUID.randomUUID().toString(),
 *     text = "Learn Jetpack Compose",
 *     completed = false,
 *     timestamp = System.currentTimeMillis()
 * )
 *
 * // Create a completed version
 * val completedTodo = todo.copy(completed = true)
 * ```
 */
@Entity(tableName = "todos")
@Serializable
@Immutable
data class Todo(
    @PrimaryKey
    val id: String,
    val text: String,
    val completed: Boolean = false,
    val timestamp: Long = System.currentTimeMillis()
) {
    companion object {
        /**
         * Creates a new Todo with a generated ID and current timestamp
         *
         * @param text The todo description
         * @param completed Initial completion state (default: false)
         * @return A new Todo instance
         */
        fun create(text: String, completed: Boolean = false): Todo {
            return Todo(
                id = java.util.UUID.randomUUID().toString(),
                text = text,
                completed = completed,
                timestamp = System.currentTimeMillis()
            )
        }
    }
}

/**
 * TodoFilter - Enum for filtering todos by completion state
 *
 * This enum is used in the UI to toggle between different views of the todo list.
 * It follows the Filter pattern commonly used in todo applications.
 *
 * @property ALL Show all todos regardless of completion state
 * @property ACTIVE Show only incomplete todos
 * @property COMPLETED Show only completed todos
 */
enum class TodoFilter {
    ALL,
    ACTIVE,
    COMPLETED;

    /**
     * Applies this filter to a list of todos
     *
     * @param todos The list to filter
     * @return Filtered list based on the filter type
     */
    fun filter(todos: List<Todo>): List<Todo> {
        return when (this) {
            ALL -> todos
            ACTIVE -> todos.filter { !it.completed }
            COMPLETED -> todos.filter { it.completed }
        }
    }

    /**
     * Gets a human-readable label for this filter
     *
     * @return Display name for the filter
     */
    fun getLabel(): String {
        return when (this) {
            ALL -> "All"
            ACTIVE -> "Active"
            COMPLETED -> "Completed"
        }
    }
}
